config/generate_lib/generic/build.sh

```diff
@@ -37,17 +37,16 @@ pushd $FW_TARGETDIR/mcu_ws >/dev/null
 		-DCMAKE_TOOLCHAIN_FILE=$TOOLCHAIN \
 		-DCMAKE_VERBOSE_MAKEFILE=ON; \
 
-    mkdir -p $FW_TARGETDIR/libmicroros; cd $FW_TARGETDIR/libmicroros; \
-	for file in $(find $FW_TARGETDIR/mcu_ws/install/lib/ -name '*.a'); do \
-		folder=$(echo $file | sed -E "s/(.+)\/(.+).a/\2/"); \
-		mkdir -p $folder; cd $folder; ar x $file; \
-		for f in *; do \
-			mv $f ../$folder-$f; \
-		done; \
-		cd ..; rm -rf $folder; \
+	mkdir -p $FW_TARGETDIR/libmicroros; cd $FW_TARGETDIR/libmicroros; \
+	echo "CREATE libmicroros.a" > $FW_TARGETDIR/libmicroros/ar_script.m; \
+	for file in $(find "$FW_TARGETDIR/install/lib/" -name '*.a'); do \
+		echo "ADDLIB ${file}" >> $FW_TARGETDIR/libmicroros/ar_script.m; \
```

Comment:
```suggestion
		folder=$(echo $file | sed -E "s/(.+)\/(.+).a/\2/"); \
		mkdir -p $folder; cd $folder; ar x $file; \
		for f in *; do \
			mv $f ../$folder-$f; \
			echo "ADDMOD ${folder}-${f}" >> $FW_TARGETDIR/libmicroros/ar_script.m; \
		done; \
		cd ..; rm -rf $folder; \
```

As you explained on https://github.com/micro-ROS/micro_ros_setup/issues/564:

> There could be a potential downside to using it, which would be that ar appears to use a last-ADDLIB-wins approach when adding libraries with object files which have the same name.

We still need to extract and rename the symbols, as we were doing on the actual method.
With this approach, we will add all the separated objects to the M-script. This will make it slower but its needed.

Reply:
So anecdotal, but even though I'm not renaming anything, I've never had any linking failures, nor problems starting resulting binaries.

Feel free to adapt this in ways you feel necessary, but at least for me, the way `ar` handles these duplicated object files has not caused any trouble so far.

Reply:
This problem arises mostly with message packages, where some packages use the same name for their structures. Example: https://github.com/micro-ROS/micro_ros_arduino/issues/1141#issuecomment-1225476670

It could also happen with custom packages added by users.

Reply:
I'll keep an eye out for this.

Just verified the symbols mentioned in https://github.com/micro-ROS/micro_ros_arduino/issues/1141#issuecomment-1225476670 all seem to be present.

Reply:
Out of curiosity, I checked, and it appears what I wrote in #564 may not actually be true.

The `libmicroros.a` resulting from the M-script contains both instances of `gripper_command__type_support_c.c.obj` for instance.

One with the symbols for the `msg/GripperCommand` (I've omitted all other symbols):

<details><summary>Click to expand</summary>

```
gripper_command__type_support.c.obj:

SYMBOL TABLE:
00000000 l    df *ABS*	00000000 gripper_command__type_support.c
...
00000000 l     O .data	00000078 control_msgs__msg__GripperCommand__rosidl_typesupport_introspection_c__GripperCommand_message_member_array
00000038 l     O .rodata	0000001c control_msgs__msg__GripperCommand__rosidl_typesupport_introspection_c__GripperCommand_message_members
00000078 l     O .data	0000000c control_msgs__msg__GripperCommand__rosidl_typesupport_introspection_c__GripperCommand_message_type_support_handle
00000000 g     F .text	00000014 control_msgs__msg__GripperCommand__rosidl_typesupport_introspection_c__GripperCommand_init_function
00000000         *UND*	00000000 control_msgs__msg__GripperCommand__init
00000014 g     F .text	00000014 control_msgs__msg__GripperCommand__rosidl_typesupport_introspection_c__GripperCommand_fini_function
00000000         *UND*	00000000 control_msgs__msg__GripperCommand__fini
00000028 g     F .text	0000001d rosidl_typesupport_introspection_c__get_message_type_support_handle__control_msgs__msg__GripperCommand
...
```

</details>

and one with the `action/GripperCommand` symbols (I've omitted all other symbols again):

<details><summary>Click to expand</summary>

```
gripper_command__type_support_c.c.obj:

SYMBOL TABLE:
00000000 l    df *ABS*	00000000 gripper_command__type_support_c.c
...
00000000 l     F .text	00000047 _GripperCommand_Goal__cdr_serialize
00000047 l     F .text	00000047 _GripperCommand_Goal__cdr_deserialize
000000d7 l     F .text	0000001c _GripperCommand_Goal__get_serialized_size
00000125 l     F .text	0000001c _GripperCommand_Goal__max_serialized_size
00000000 l     O .data	0000001c __callbacks_GripperCommand_Goal
0000001c l     O .data	0000000c _GripperCommand_Goal__type_support
0000014b l     F .text	00000091 _GripperCommand_Result__cdr_serialize
000001dc l     F .text	00000087 _GripperCommand_Result__cdr_deserialize
00000316 l     F .text	0000001c _GripperCommand_Result__get_serialized_size
000003bf l     F .text	0000001c _GripperCommand_Result__max_serialized_size
00000028 l     O .data	0000001c __callbacks_GripperCommand_Result
00000044 l     O .data	0000000c _GripperCommand_Result__type_support
000003e5 l     F .text	00000091 _GripperCommand_Feedback__cdr_serialize
00000476 l     F .text	00000087 _GripperCommand_Feedback__cdr_deserialize
000005b0 l     F .text	0000001c _GripperCommand_Feedback__get_serialized_size
00000659 l     F .text	0000001c _GripperCommand_Feedback__max_serialized_size
00000050 l     O .data	0000001c __callbacks_GripperCommand_Feedback
0000006c l     O .data	0000000c _GripperCommand_Feedback__type_support
0000067f l     F .text	00000067 _GripperCommand_SendGoal_Request__cdr_serialize
000006e6 l     F .text	00000067 _GripperCommand_SendGoal_Request__cdr_deserialize
000007ae l     F .text	0000001c _GripperCommand_SendGoal_Request__get_serialized_size
00000811 l     F .text	0000001c _GripperCommand_SendGoal_Request__max_serialized_size
00000078 l     O .data	0000001c __callbacks_GripperCommand_SendGoal_Request
00000094 l     O .data	0000000c _GripperCommand_SendGoal_Request__type_support
00000837 l     F .text	00000065 _GripperCommand_SendGoal_Response__cdr_serialize
0000089c l     F .text	0000005f _GripperCommand_SendGoal_Response__cdr_deserialize
00000966 l     F .text	0000001c _GripperCommand_SendGoal_Response__get_serialized_size
000009d0 l     F .text	0000001c _GripperCommand_SendGoal_Response__max_serialized_size
000000a0 l     O .data	0000001c __callbacks_GripperCommand_SendGoal_Response
000000bc l     O .data	0000000c _GripperCommand_SendGoal_Response__type_support
000000c8 l     O .data	00000010 GripperCommand_SendGoal__callbacks
000000d8 l     O .data	0000000c GripperCommand_SendGoal__handle
00000a00 l     F .text	00000047 _GripperCommand_GetResult_Request__cdr_serialize
00000a47 l     F .text	00000047 _GripperCommand_GetResult_Request__cdr_deserialize
00000ad7 l     F .text	0000001c _GripperCommand_GetResult_Request__get_serialized_size
00000b25 l     F .text	0000001c _GripperCommand_GetResult_Request__max_serialized_size
000000e4 l     O .data	0000001c __callbacks_GripperCommand_GetResult_Request
00000100 l     O .data	0000000c _GripperCommand_GetResult_Request__type_support
00000b4b l     F .text	00000065 _GripperCommand_GetResult_Response__cdr_serialize
00000bb0 l     F .text	0000005f _GripperCommand_GetResult_Response__cdr_deserialize
00000c7a l     F .text	0000001c _GripperCommand_GetResult_Response__get_serialized_size
00000ce4 l     F .text	0000001c _GripperCommand_GetResult_Response__max_serialized_size
0000010c l     O .data	0000001c __callbacks_GripperCommand_GetResult_Response
00000128 l     O .data	0000000c _GripperCommand_GetResult_Response__type_support
00000134 l     O .data	00000010 GripperCommand_GetResult__callbacks
00000144 l     O .data	0000000c GripperCommand_GetResult__handle
00000d14 l     F .text	00000067 _GripperCommand_FeedbackMessage__cdr_serialize
00000d7b l     F .text	00000067 _GripperCommand_FeedbackMessage__cdr_deserialize
00000e43 l     F .text	0000001c _GripperCommand_FeedbackMessage__get_serialized_size
00000ea6 l     F .text	0000001c _GripperCommand_FeedbackMessage__max_serialized_size
00000150 l     O .data	0000001c __callbacks_GripperCommand_FeedbackMessage
0000016c l     O .data	0000000c _GripperCommand_FeedbackMessage__type_support
00000000         *UND*	00000000 rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__msg__GripperCommand
0000008e g     F .text	00000049 get_serialized_size_control_msgs__action__GripperCommand_Goal
00000000         *UND*	00000000 get_serialized_size_control_msgs__msg__GripperCommand
000000f3 g     F .text	00000032 max_serialized_size_control_msgs__action__GripperCommand_Goal
00000000         *UND*	00000000 max_serialized_size_control_msgs__msg__GripperCommand
00000141 g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_Goal
00000263 g     F .text	000000b3 get_serialized_size_control_msgs__action__GripperCommand_Result
00000332 g     F .text	0000008d max_serialized_size_control_msgs__action__GripperCommand_Result
000003db g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_Result
000004fd g     F .text	000000b3 get_serialized_size_control_msgs__action__GripperCommand_Feedback
000005cc g     F .text	0000008d max_serialized_size_control_msgs__action__GripperCommand_Feedback
00000675 g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_Feedback
0000074d g     F .text	00000061 get_serialized_size_control_msgs__action__GripperCommand_SendGoal_Request
000007ca g     F .text	00000047 max_serialized_size_control_msgs__action__GripperCommand_SendGoal_Request
0000082d g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_SendGoal_Request
000008fb g     F .text	0000006b get_serialized_size_control_msgs__action__GripperCommand_SendGoal_Response
00000982 g     F .text	0000004e max_serialized_size_control_msgs__action__GripperCommand_SendGoal_Response
000009ec g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_SendGoal_Response
000009f6 g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_service_type_support_handle__control_msgs__action__GripperCommand_SendGoal
00000a8e g     F .text	00000049 get_serialized_size_control_msgs__action__GripperCommand_GetResult_Request
00000af3 g     F .text	00000032 max_serialized_size_control_msgs__action__GripperCommand_GetResult_Request
00000b41 g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_GetResult_Request
00000c0f g     F .text	0000006b get_serialized_size_control_msgs__action__GripperCommand_GetResult_Response
00000c96 g     F .text	0000004e max_serialized_size_control_msgs__action__GripperCommand_GetResult_Response
00000d00 g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_GetResult_Response
00000d0a g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_service_type_support_handle__control_msgs__action__GripperCommand_GetResult
00000de2 g     F .text	00000061 get_serialized_size_control_msgs__action__GripperCommand_FeedbackMessage
00000e5f g     F .text	00000047 max_serialized_size_control_msgs__action__GripperCommand_FeedbackMessage
00000ec2 g     F .text	0000000a rosidl_typesupport_microxrcedds_c__get_message_type_support_handle__control_msgs__action__GripperCommand_FeedbackMessage
...
```

</details>

How it exactly works I'm not sure, but the linker seems to be able to find all of them when needed.

This is without renaming anything.

I can't found the source of my "last-`ADDLIB` wins", so perhaps it's actually not true.

Regardless: no need to keep this closed of course.

Reply:
Let us test this in our side

Reply:
Apologies, that last sentence should have read:

> No need to re-open this of course.

I did not mean to imply this should be re-opened.